In [ ]:
!pip install dlt[clickhouse]

In [2]:
from pyiceberg.catalog import load_catalog
import pyarrow as pa
import dlt
import pandas as pd
import pyarrow.dataset as ds
import pyarrow.fs as fs
from dlt.sources.filesystem import filesystem

## **Connect to the Nessie catalog**

Re-establish the connection to the **Nessie REST catalog**.  
This allows us to read metadata and schema information from the Iceberg tables  
that were created and loaded in the previous steps.

We verify the connection by listing all available namespaces.


In [ ]:
# Configure the connection to the Nessie REST catalog
catalog = load_catalog(
    "nessie",
    **{
        "uri": "http://nessie:19120/iceberg/main/",
    }
)

# Verify the connection by listing the namespaces
namespaces = catalog.list_namespaces()
print("Namespaces:", namespaces)

Namespaces: [('taxis-project',)]


## **Extract Iceberg data and load it into ClickHouse**

This notebook executes the **fourth and final stage** of the project pipeline:  
**migrating the Iceberg data stored in MinIO into a ClickHouse database** for analytical querying.

**Process overview:**

1. **Connect to MinIO**  
   - Configure an S3-compatible connection to MinIO using access credentials.

2. **Define the DLT resource (`iceberg_df`)**  
   - Load the Iceberg table `taxis-project.taxis` from the Nessie catalog.  
   - Scan the table and convert it into an Arrow Table.  
   - Yield the data in batches, allowing efficient ingestion into ClickHouse.

3. **Create and configure the pipeline**  
   - Name: `clickhouse_pipeline`  
   - Destination: `clickhouse`  
   - Dataset name: `taxis_project`.

4. **Execute the pipeline**  
   - Runs the DLT job that transfers all Iceberg table data into ClickHouse.  
   - The parameter `write_disposition="replace"` ensures a full table refresh.


In [ ]:
s3 = fs.S3FileSystem(
    endpoint_override="http://minio:9000",  # inside Docker: use "minio:9000", from local: "localhost:9000"
    access_key="admin",
    secret_key="password",
    region="us-east-1"
)

iceberg_table_path = "my-bucket/taxis-project/taxis"

@dlt.resource(table_name="taxis")
def iceberg_df():
    # Load the Iceberg table from the Nessie catalog
    taxis = catalog.load_table("taxis-project.taxis")
    
    # Run the table scan and get an Arrow Table
    arrow_table = taxis.scan().to_arrow()
    
    # Iterate through Arrow batches
    for batch in arrow_table.to_batches():
        yield batch


# Initialize the DLT pipeline
pipeline = dlt.pipeline(
  pipeline_name='clickhouse_pipeline',
  destination='clickhouse',
  dataset_name='taxis_project'
)

# Define the source (filesystem connector)
source = filesystem()

# Run the pipeline
load_info = pipeline.run(
    iceberg_df,             # use the Iceberg table resource
    loader_file_format="parquet",
    write_disposition="replace"
)
print(load_info)

2025-10-21 14:59:20,618|[WARNING]|1153|138890886543168|dlt|type_mapping.py|to_db_datetime_type:58|Column flags for timezone or precision are not yet supported in this destination. One or both of these flags were used in column 'tpep_pickup_datetime'. in table 'taxis'.
2025-10-21 14:59:20,618|[WARNING]|1153|138890886543168|dlt|type_mapping.py|to_db_datetime_type:58|Column flags for timezone or precision are not yet supported in this destination. One or both of these flags were used in column 'tpep_dropoff_datetime'. in table 'taxis'.
2025-10-21 14:59:20,662|[WARNING]|1153|138890886543168|dlt|type_mapping.py|to_db_datetime_type:58|Column flags for timezone or precision are not yet supported in this destination. One or both of these flags were used in column 'tpep_pickup_datetime'. in table 'taxis'.
2025-10-21 14:59:20,663|[WARNING]|1153|138890886543168|dlt|type_mapping.py|to_db_datetime_type:58|Column flags for timezone or precision are not yet supported in this destination. One or both 

Pipeline clickhouse_pipeline load step completed in 2.91 seconds
1 load package(s) were loaded to destination clickhouse and into dataset taxis_project
The clickhouse destination used clickhouse://default:***@clickhouse:9000/default location to store data
Load package 1761058758.772736 is LOADED and contains no failed jobs


The pipeline successfully loaded the Iceberg data into ClickHouse in approximately **2.9 seconds**.

- Destination: ClickHouse database (`taxis_project` dataset).  
- Loader format: Parquet.  
- Disposition: Replace (existing table overwritten).  
- No failed jobs detected.

Once this step is verified with record counts and metadata queries,
the full end-to-end data pipeline is successfully implemented.
